### PR TITLE
Extend HasProcessingInformation and HasInstallationInformation

### PR DESCRIPTION
**Pull request by takebene**
## Pull Request

- [x] I have checked for similar PRs.
- [x] I have read the [contributing guidelines](https://github.com/4Soft-de/kbl-model/blob/develop/.github/CONTRIBUTING.md).

### Changes

- [x] Code
- [ ] Documentation
- [ ] Other: 

### Description

Extend HasProcessingInformation and HasInstallationInformation interfaces to filter Instruction

### CODE REVIEW

**Comment by RainerGanss** [comment]
btw, cool branch name ;-)

File: v24/src/main/java/com/foursoft/harness/kbl/v24/util/StreamUtils.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    /**
+     * Returns an optional if the list was empty or contained one element.
+     * <p>
+     * If the list contains more than one element, a log-error statement is
+     * created with the calling stacktrace attached. In this case the first
+     * element is taken.
+     */
+    public static <T> Collector<T, List<T>, Optional<T>> findOneOrNone() {
```
[the remark this answers is not in the document]
Comment (takebene) [reply v24/src/main/java/com/foursoft/harness/kbl/v24/util/StreamUtils.java:54]: Can we do this if we merge all the Harness projects?

File: v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java [hunk 2]
```diff
@@ -25,9 +25,37 @@
...
+     *
+     * @param matches defines the meaning of the value
+     * @return the first value with the given type.
+     */
+    default Optional<String> getInstallationInstructionValue(final Predicate<KblInstallationInstruction> matches) {
+        return getInstallationInformations()
+                .stream()
+                .filter(matches)
+                .map(KblInstallationInstruction::getInstructionValue)
+                .collect(StreamUtils.findOneOrNone());
```
[the remark this answers is not in the document]
Comment (takebene) [reply v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java:59]: hmm, then I will adapt the documentation and create another method that returns a list. That way you can decide for yourself.

File: v24/src/main/java/com/foursoft/harness/kbl/v24/HasProcessingInformation.java [hunk 3]
```diff
@@ -25,10 +25,58 @@
...
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
 
 public interface HasProcessingInformation {
 
     List<KblProcessingInstruction> getProcessingInformations();
 
+    /**
+     * Filters the list of {@link KblProcessingInstruction} key.
```
Comment (RainerGanss) [inline v24/src/main/java/com/foursoft/harness/kbl/v24/HasProcessingInformation.java:40]: Please add to those java docs that at most one element of the given typ is expected. If more a found, the first one will be returned and a warning is logged. 
If multiple values of the given type are valid, than please use the other #getProcessingInstructionValues(...) method.

File: v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    @Test
+    void processingInformationTest() throws IOException {
+        try (final InputStream is = getClass().getClassLoader().getResourceAsStream("sample.kbl")) {
+            final KBLContainer kblContainer = new KblReader().read(is);
+
+            final KblSlotOccurrence occurrence = kblContainer.getHarness().getConnectorOccurrences()
+                    .stream()
+                    .filter(c -> c.getId().equals("Id184"))
+                    .flatMap(c -> c.getSlots().stream())
+                    .findFirst().orElse(null);
```
Comment (DManstrator) [inline v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java:49]: Just for consistency: Why not also using `StreamUtils.findOne` like in the other test?

**Review by DManstrator — CHANGES_REQUESTED** [review]

File: v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java (hunk 4 above)
Comment (takebene) [reply v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java:49]: no log information needed

File: v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java [hunk 5]
```diff
@@ -25,9 +25,58 @@
...
+
+    /**
+     * Filters the list of {@link KblInstallationInstruction} key.
+     *
+     * @param instructionType defines the meaning of the value
+     * @return a stream with all instruction values for the given type.
+     */
+    default Stream<String> getInstallationInstructionValues(final String instructionType) {
+        return getInstallationInstructionValues(c -> c.getInstructionType().equalsIgnoreCase(instructionType));
+    }
```
[the remark this answers is not in the document]
Comment (takebene) [reply v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java:72]: i don't like any information about the return value in the method name. 
I renamed the method anyway. This is how it is often done: java.nio.file.File.lines()

[the remark this answers is not in the document]
Comment (DManstrator) [reply v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java:72]: Yeah, I can totally see that point. I just had a problem with the "get" in the method name. In your example, it's also not present, `installationInstructionValues` would also be okay.

File: v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+            assertThat(occurrence)
+                    .returns(Optional.of("Instruction_value396"), c -> c.getProcessingInstructionValue("Instruction_type396"))
+                    .returns(List.of("Instruction_value396"), c -> c.getProcessingInstructionValues("Instruction_type396")
```
Comment (RainerGanss) [inline v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java:53]: Can you please add the Predicate methods also here so sonar is happy with the test coverage?

File: v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java [hunk 7]
```diff
@@ -25,9 +25,58 @@
...
+import java.util.stream.Stream;
 
 public interface HasInstallationInformation {
 
     List<KblInstallationInstruction> getInstallationInformations();
+
+    /**
+     * Filters the list of {@link KblInstallationInstruction} key.
+     *
+     * @param instructionType defines the meaning of the value
```
[the remark this answers is not in the document]
Comment (takebene) [reply v24/src/main/java/com/foursoft/harness/kbl/v24/HasInstallationInformation.java:44]: I think that is sufficient. Key Values are not that difficult to understand now. Besides, we have that also so with the CustomProperties

File: v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java (hunk 6 above)
Comment (takebene) [reply v24/src/test/java/com/foursoft/harness/kbl/v24/HasProcessingInformationTest.java:53]: Sonar says 100% coverage only with the StreamUtils the coverage is not so good

**Review by RainerGanss — APPROVED** [review]

**Pull request merged**